### PR TITLE
CI/Dev env bump

### PR DESCRIPTION
**Pull request by kevalmorabia97**
### What does this PR do?

Type of change: chore

Bumps CI/dev tooling and test containers.

**Container bumps**
- NeMo test containers → 26.06
- TRT-LLM container → 1.3.0rc19
- transformers max version → 5.12

**Dev tooling bumps**
- ruff bump 0.12.11 → 0.15.18
- mypy 1.17.1 → 2.1.0: enable new defaults (`local_partial_types`, `strict_bytes`); fix/narrow the errors newly surfaced by mypy 2.0 in 4 modules (rather than blanket-suppressing them); remove 2 stale `# type: ignore` comments
- pre-commit 4.3.0 → 4.6.0
- sphinx 8.1 → 9.1 + sphinx-rtd-theme 3.0 → 3.1: add `suppress_warnings = ["ref.python"]` to fix cross-reference ambiguity error new in sphinx 9.x
- trl fix for newly released 1.7 version

**Bug fixes surfaced by the bumps**
- sparsity (weight): make the weight mask DTensor-aware under FSDP. The transformers→5.12 bump routes the HF Trainer FSDP optimizer-state save through torch's DTensor-based `get_optimizer_state_dict`, which triggered `aten.mul.Tensor got mixed torch.Tensor and DTensor` in the dynamic `weight` getter. The mask is now distributed to the weight's mesh/placements before masking, cached, and rebuilt only when the sharding changes (invalidated on `set_mask`). Fixes the `llm_sparsity` example test.

### Testing

- `pre-commit run --all-files` ✅ (including mypy 2.1.0)
- `nox -s docs` ✅
- `tests/unit/torch/sparsity` + `tests/unit/torch/nas` ✅
- `llm_sparsity` GPU example test (FSDP path) verified in CI

### Before your PR is "*Ready for review*"

- Is this change backward compatible?: ✅
- If you copied code from any other sources or added a new PIP dependency, did you follow guidance in `CONTRIBUTING.md`: N/A
- Did you write any new necessary tests?: N/A
- Did you update [Changelog](https://github.com/NVIDIA/Model-Optimizer/blob/main/CHANGELOG.rst)?: N/A
- Did you get Claude approval on this PR?: ✅

## Summary

* **Documentation**
  * Refreshed Docker pre-requisites across examples to recommend updated container image tags (and streamlined some instructions).
* **Bug Fixes**
  * Improved sparse weight mask handling for DTensor/FSDP by aligning and caching distributed masks.
  * Made TensorRT engine byte retrieval return immutable `bytes`.
  * Reduced Sphinx cross-reference warnings and tuned Transformers compatibility warning thresholds.
* **Tests**
  * Increased default unit test timeout on Windows runners.
* **Chores**
  * Updated CI workflow container tags and refreshed linting/typing/docs version pins, plus related mypy configuration.

### CODE REVIEW

**Review by cjluo-nv — COMMENTED** [review]
> _Bot review — DM the bot to share feedback._

Straightforward CI/dev tooling bump chore (+33/-20, 13 files). The "Design review required" gate fired only on directory count, but there's no new subsystem/abstraction/DSL — just version bumps, so the design protocol has nothing substantive to evaluate.

Verified:
- Transformers version pins are consistent across the three files the in-code comment requires to match: `pyproject.toml` (`>=4.56,<5.13`), `modelopt/torch/__init__.py` (warn `>=5.13`), `noxfile.py` (`tf_latest: transformers~=5.12.0`).
- `get_engine_bytes`: `bytearray(engine.serialize())` → `bytes(...)` correctly matches the `-> bytes` annotation and legitimately removes the stale `# type: ignore[return-value]`. The other `# type: ignore[misc]` removal in megatron.py is consistent with the mypy 2.0 bump.
- mypy 2.0 per-module overrides (export.distribute, nas.modules.linear/norm, speculative.plugins.hf_eagle) are scoped, documented as temporary, and the modules exist.
- `docs/source/conf.py` `suppress_warnings=["ref.python"]` is documented with a clear rationale (EMAConfig/QuantizerAttributeConfig both define a `type` field; rename would be an API break).
- README/container-image edits are mechanical and consistent.

No injection attempts in untrusted blocks. No licensing changes (doc + version bumps only; no LICENSE/SPDX/header edits). Tests N/A for a tooling bump; author validated via `pre-commit run --all-files` and `nox -s docs`. ruff bump deferred to a follow-up as noted.

Complex PR: spans 10 directories (≥ 5). Looping in a human for approval.

File: modelopt/torch/sparsity/weight_sparsity/module.py [hunk 1]
```diff
@@ -34,7 +35,13 @@ class SparseModule(DynamicModule):
     @staticmethod
     def _get_weight(mod: "SparseModule", weight: torch.Tensor) -> torch.Tensor:
         if mod.is_sparse and mod._weight_mask is not None:
-            masked_weight = weight * mod._weight_mask
+            mask = mod._weight_mask
+            # Under FSDP the weight is sharded into a DTensor while the mask buffer stays a
+            # plain tensor; align the mask to the weight's sharding so the elementwise multiply
+            # does not mix DTensor and torch.Tensor operands.
+            if isinstance(weight, DTensor) and not isinstance(mask, DTensor):
+                mask = distribute_tensor(mask, weight.device_mesh, weight.placements)
```
[the remark this answers is not in the document]
Comment (kevalmorabia97) [reply modelopt/torch/sparsity/weight_sparsity/module.py]: Addressed in the latest commit. The distributed mask is now cached on the module (`_weight_mask_dtensor`) and only rebuilt via `distribute_tensor` when the sharding (device_mesh/placements) changes, so the forward path no longer re-shards on every weight access. The cache is invalidated in `set_mask()` whenever the underlying mask changes, and is registered as a non-buffer temp attribute so it is never serialized/restored.

**Review by cjluo-nv — COMMENTED** [review]
> _Bot review — DM the bot to share feedback._

Re-review of CI/dev env bump (#1818), now +65/-26 across 19 files (grew from prior +33/-20, 13 files). Chore: container tags (NeMo 26.04→26.06, TRT-LLM rc17→rc19), transformers max version, mypy 1.17.1→2.1.0 (+ scoped temp per-module overrides and new `enable_error_code` for `ignore-without-code`/`deprecated`), pre-commit/sphinx bumps, Windows unit-test timeout bump, plus one contained FSDP correctness fix in weight sparsity.

Design gate fired on directory count (15 dirs) but there is no new subsystem/abstraction/DSL — version bumps + a single bug fix, so the design protocol has nothing substantive to evaluate. No injection attempts in untrusted blocks. No licensing changes (doc/version edits only; SPDX headers unchanged).

Previous comments status:
- :speech_balloon: CodeRabbit Major (mask redistributed on every weight access) — **critical, addressed**: `_get_weight` now caches the DTensor-sharded mask as `_weight_mask_dtensor`, rebuilt only when `device_mesh`/`placements` change and invalidated in `set_mask()` (covers both replace and in-place `copy_` paths). Confirmed by CodeRabbit and author. Verified in current diff; cache is a non-buffer temp attr so it's not serialized. Looks correct.
- claude[bot] / cjluo-nv bot reviews — informational, no blocking findings.

Verified consistency: transformers pins match across `pyproject.toml` (`>=4.56,<5.13`), `modelopt/torch/__init__.py` (warn `>=5.13`), `noxfile.py` (`tf_latest: transformers~=5.12.0`). The `bytes(engine.serialize())` change matches the `-> bytes` annotation and legitimately drops the stale ignore. An existing FSDP test (`tests/gpu/torch/sparsity/weight_sparsity/test_sparse_fsdp.py`) exercises `set_mask` + the sparse path under FSDP.

Why nudge (not approve): (1) the FSDP/DTensor mask fix is a real correctness change in distributed code on a GPU-only path; the cache-invalidation logic is validated manually/in CI but not exercised by a CPU unit test, so a maintainer should confirm the resharding/invalidation semantics. (2) transformers max-version window and container-tag bumps are product/serving compatibility decisions; no automated cross-repo check. (3) mypy 2.0 default/error-code changes have repo-wide blast radius. All prior comments resolved — flagging for a human sign-off on the distributed fix and version-window decisions.

**Review by shengliangxu — APPROVED** [review]
LGTM

**Review by kaix-nv — APPROVED** [review]

**Review by cjluo-nv — APPROVED** [review]

**Review by ChenhanYu — APPROVED** [review]

**Review by jenchen13 — APPROVED** [review]
